src/hp/derp/client.rs

```diff
@@ -0,0 +1,515 @@
+use std::time::Duration;
+use std::{net::SocketAddr, sync::Arc};
+
+use anyhow::{bail, Context, Result};
+use bytes::BytesMut;
+use governor::RateLimiter;
+use postcard::experimental::max_size::MaxSize;
+use quinn::AsyncUdpSocket;
+use serde::{Deserialize, Serialize};
+use tokio::{
+    io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt},
+    sync::Mutex,
+};
+
+use super::{
+    read_frame, read_frame_header, server::ServerInfo, write_frame, write_frame_header, FrameType,
+    FRAME_CLIENT_INFO, FRAME_CLOSE_PEER, FRAME_FORWARD_PACKET, FRAME_HEALTH, FRAME_NOTE_PREFERRED,
+    FRAME_PEER_GONE, FRAME_PING, FRAME_PONG, FRAME_RESTARTING, FRAME_SEND_PACKET,
+    FRAME_SERVER_INFO, FRAME_WATCH_CONNS, MAGIC, MAX_PACKET_SIZE, PROTOCOL_VERSION,
+};
+use crate::hp::{
+    derp::{
+        FRAME_KEEP_ALIVE, FRAME_PEER_PRESENT, FRAME_RECV_PACKET, FRAME_SERVER_KEY, MAX_FRAME_SIZE,
+        MAX_INFO_LEN, NONCE_LEN,
+    },
+    key::{self, node::PublicKey},
+    magicsock::Conn,
+};
+
+const SERVER_KEY_FRAME_MAX_SIZE: u32 = 1024;
+
+/// A DERP Client.
+struct Client<W, R, S, C, MW = governor::middleware::NoOpMiddleware>
+where
+    W: AsyncWrite + Send + Unpin + 'static, // TODO: static?
+    R: AsyncRead + Unpin,
+    S: governor::state::DirectStateStore<Key = governor::state::direct::NotKeyed>,
+    C: governor::clock::Clock,
+    MW: governor::middleware::RateLimitingMiddleware<C::Instant>,
+{
+    /// Server key of the DERP server, not a machine or node key
+    server_key: key::node::PublicKey,
+    /// TODO: maybe change to "secret_key" to match `key::node::SecretKey` naming
+    private_key: key::node::SecretKey,
+    /// public key associated with `private_key`
+    public_key: key::node::PublicKey,
+    conn: Conn,
+    reader: R,
+    /// TODO: This is a string in the go impl, using bytes here to make it easier for postcard
+    /// to serialize. 32 is a random number I chose. Need to figure out what the `mesh_key`
+    /// is in practice.
+    mesh_key: [u8; 32],
+    can_ack_pings: bool,
+    is_prober: bool,
+
+    // mutex lock to protect the writer
+    writer: Arc<Mutex<W>>,
+    // TODO: maybe write a trait to make working with the rate limiter less gross cause it's currently disgusting
+    rate_limiter: Option<Arc<RateLimiter<governor::state::direct::NotKeyed, S, C, MW>>>,
+    /// sticky (set by Recv)
+    /// TODO: temporarily a string until I figure out what to do about cloning errors
+    recv_err: Arc<Mutex<Option<String>>>,
+}
+
+/// TODO: ClientBuilder
+
+impl<W, R, S, C, MW> Client<W, R, S, C, MW>
+where
+    W: AsyncWrite + Send + Unpin + 'static,
+    R: AsyncRead + Unpin,
+    S: governor::state::DirectStateStore<Key = governor::state::direct::NotKeyed>,
+    C: governor::clock::Clock,
+    MW: governor::middleware::RateLimitingMiddleware<C::Instant>,
+{
+    // TODO: for something relatively straight forward, this is pretty hard to follow
+    // TODO: also, should Client.server_key be an option?
```

Comment:
I'm just going to leave this for now because I think part of the problem is the original design for `read_frame` that the go impl has, and all the reading / writing needs to be looked at more carefully anyway. But thank you for dropping some knowledge here.